File: be/app/schemas/food.py
from typing import List, Optional

from pydantic import BaseModel

from app.schemas.base import BaseSchema, TimestampMixin


class FoodCategoryBase(BaseModel):
    name: str
    status: bool = True


class FoodCategoryCreate(FoodCategoryBase):
    pass


class FoodCategoryUpdate(BaseModel):
    name: Optional[str] = None
    status: Optional[bool] = None


class FoodCategory(FoodCategoryBase, TimestampMixin, BaseSchema):
    id: int


class FoodBase(BaseModel):
    name: str
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    country_id: Optional[int] = None
    region_id: Optional[int] = None
    district_id: Optional[int] = None
    ward_id: Optional[int] = None
    category_id: int
    thumbnail_url: Optional[str] = None
    price_min: Optional[float] = None
    price_max: Optional[float] = None
    status: bool = True


class FoodCreate(FoodBase):
    pass


class FoodUpdate(BaseModel):
    name: Optional[str] = None
    name_code: Optional[str] = None
    description: Optional[str] = None
    description_code: Optional[str] = None
    country_id: Optional[int] = None
    region_id: Optional[int] = None
    district_id: Optional[int] = None
    ward_id: Optional[int] = None
    category_id: Optional[int] = None
    thumbnail_url: Optional[str] = None
    price_min: Optional[float] = None
    price_max: Optional[float] = None
    popularity_score: Optional[float] = None
    status: Optional[bool] = None


class Food(FoodBase, TimestampMixin, BaseSchema):
    id: int
    popularity_score: float = 0


class RatingBase(BaseModel):
    reference_id: int
    reference_type: str
    user_id: int
    rating: float
    comment: Optional[str] = None


class RatingCreate(RatingBase):
    pass


class RatingUpdate(BaseModel):
    rating: Optional[float] = None
    comment: Optional[str] = None


class Rating(RatingBase, TimestampMixin, BaseSchema):
    id: int

File: java-springboot/src/main/java/com/william/travel/entity/Food.java
package com.william.travel.entity;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import lombok.EqualsAndHashCode;
import lombok.ToString;

import java.math.BigDecimal;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "food")
@Getter
@Setter
@EqualsAndHashCode(callSuper = true, exclude = {"country", "region", "district", "ward", "category", "ratings", "media"})
@ToString(exclude = {"country", "region", "district", "ward", "category", "ratings", "media"})
public class Food extends BaseEntity {

    @Column(name = "name", length = 255, nullable = false)
    private String name;

    @Column(name = "name_code", length = 255)
    private String nameCode;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "description_code", columnDefinition = "TEXT")
    private String descriptionCode;

    @Column(name = "thumbnail_url", length = 255)
    private String thumbnailUrl;

    @Column(name = "price_min", precision = 10, scale = 2)
    private BigDecimal priceMin;

    @Column(name = "price_max", precision = 10, scale = 2)
    private BigDecimal priceMax;

    @Column(name = "popularity_score")
    private Double popularityScore = 0.0;

    @Column(name = "status", nullable = false)
    private Boolean status = true;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "country_id")
    private Country country;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "region_id")
    private Region region;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "district_id")
    private District district;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "ward_id")
    private Ward ward;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private FoodCategory category;

    @OneToMany(mappedBy = "food", fetch = FetchType.LAZY)
    private Set<Rating> ratings = new HashSet<>();

    @OneToMany(mappedBy = "food", fetch = FetchType.LAZY)
    private Set<Media> media = new HashSet<>();
}
